File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int s_length = s.size();
        
        int max_length = 0;
        int max_length_idx = -1;

        auto expand = [&](int left, int right) {
            while (left >= 0 && right < s.size() && s[left] == s[right]) {
                left--;
                right++;
            }
            int curr_length = right - left - 1;
            if (curr_length > max_length) {
                max_length = curr_length;
                max_length_idx = left + 1;
            }
        };

        for (int i = 0; i < s_length; i++) {
            expand(i, i);
        }
        
        for (int i = 0; i < s_length - 1; i++) {
            expand(i, i + 1);
        }

        return s.substr(max_length_idx, max_length);
    }
};

// babad
// i = 1
// getPalindrome(1, 1)
//  left = 1, right = 1
//  left = 0, right = 2
//  left = -1, right = 3
//  start at 0; length = 3

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expand(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return left + 1, right

        max_str_left = -1
        max_str_right = -1

        for i in range(len(s)):
            str_left, str_right = expand(i, i)
            if str_right - str_left > max_str_right - max_str_left:
                max_str_right = str_right
                max_str_left = str_left
        
        for i in range(len(s) - 1):
            str_left, str_right = expand(i, i + 1)
            if str_right - str_left > max_str_right - max_str_left:
                max_str_right = str_right
                max_str_left = str_left
        
        return s[max_str_left:max_str_right]